sockeye/inference.py

```diff
@@ -376,33 +388,45 @@ def load_models(context: mx.context.Context,
     :param cache_output_layer_w_b: Models cache weights and biases for logit computation as NumPy arrays (used with
                                    restrict lexicon).
-    :return: List of models, source vocabulary, target vocabulary.
+    :return: List of models, source vocabulary, target vocabulary, source factor vocabularies.
     """
-    models, source_vocabs, target_vocabs = [], [], []
+    models = []  # type: List[InferenceModel]
+    source_vocabs = []  # type: List[List[vocab.Vocab]]
+    target_vocabs = []  # type: List[vocab.Vocab]
+
     if checkpoints is None:
         checkpoints = [None] * len(model_folders)
+
     for model_folder, checkpoint in zip(model_folders, checkpoints):
-        source_vocabs.append(vocab.vocab_from_json_or_pickle(os.path.join(model_folder, C.VOCAB_SRC_NAME)))
-        target_vocabs.append(vocab.vocab_from_json_or_pickle(os.path.join(model_folder, C.VOCAB_TRG_NAME)))
-        model = InferenceModel(model_folder=model_folder,
-                               context=context,
-                               beam_size=beam_size,
-                               batch_size=batch_size,
-                               softmax_temperature=softmax_temperature,
-                               checkpoint=checkpoint,
-                               decoder_return_logit_inputs=decoder_return_logit_inputs,
-                               cache_output_layer_w_b=cache_output_layer_w_b)
-        models.append(model)
-
-    utils.check_condition(vocab.are_identical(*source_vocabs), "Source vocabulary ids do not match")
+        model_source_vocabs = vocab.load_source_vocabs(model_folder)
+        source_vocabs.append(model_source_vocabs)
+        target_vocabs.append(vocab.vocab_from_json(os.path.join(model_folder, C.VOCAB_TRG_NAME)))
+
+        inference_model = InferenceModel(model_folder=model_folder,
+                                         context=context,
+                                         beam_size=beam_size,
+                                         batch_size=batch_size,
+                                         softmax_temperature=softmax_temperature,
+                                         checkpoint=checkpoint,
+                                         decoder_return_logit_inputs=decoder_return_logit_inputs,
+                                         cache_output_layer_w_b=cache_output_layer_w_b)
+        utils.check_condition(inference_model.num_source_factors == len(model_source_vocabs),
+                              "Number of loaded source vocabularies (%d) does not match "
+                              "number of source factors for model '%s' (%d)" % (len(model_source_vocabs), model_folder,
+                                                                                inference_model.num_source_factors))
+        models.append(inference_model)
+
     utils.check_condition(vocab.are_identical(*target_vocabs), "Target vocabulary ids do not match")
+    for fi in range(len(source_vocabs[0])):
```

Comment:
maybe for clarity source_vocabs[0] could be pulled out to:
```
first_model_vocabs = source_vocabs[0]
```

Reply:
done